### ISSUE TITLE
Add support for AAD auth to Azure Storage

### ISSUE THREAD

**Issue by dylan-smith** [github/gh-gei#673]
When migrating from GHES -> GHEC we use Azure Storage as an intermediate storage location for the migration archives. The way we authenticate to Azure Storage today is we require the user to provide an Azure Storage Connection String which includes a shared key.

Some customers would prefer that we use an AAD identity to authenticate instead of a shared key to align with their security policies. We should be able to add support to the CLI to accept a AAD service principal client ID and secret and use that to auth to the storage account.

**Comment by timrogers** [comment]
See request from Santander [here](https://github.com/githubcustomers/Santander/issues/205#issuecomment-1271712328).

**Comment by CSMonkee** [comment]
@dylan-smith - Are you able to provide an update regarding this possible CLI change please? Santander security are looking for a solution akin to that shown [here](https://github.com/githubcustomers/Santander/issues/205#issuecomment-1280464728).

**Comment by timrogers** [comment]
@CSMonkee Thanks for the follow up on this! We don't have immediate plans to add support for AAD authentication in the CLI. We recognise the value, but we don't have capacity right now. Is that what the customer is looking for still? I'm not sure I understand the diagram.

If it isn't going to be possible for the customer to use a shared key, as is supported today, then there is a possible workaround where the customer generates the migration archives manually with `ghe-migrator`, uploads them to blob storage as they wish and then uses the CLI to actually start the migration. If that might work, I'd be happy to share more details.

**Comment by danielmeppiel** [comment]
@timrogers that workaround is very promising and Santander is interested to learn more (had a call with them on 8th November and confirmed this point). Can we set up a call including Santander for us to go through how that workaround may work? Alternatively, can you share with me the technical details on how this may work, as I understand the commands are not documented? 

Really appreciated!

**Comment by danielmeppiel** [comment]
@timrogers , does that mean that using a [User Delegation SAS](https://learn.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas) would be technically feasible? In other words:

1. [Generate the migration archives manually](https://docs.github.com/en/enterprise-server@3.4/admin/user-management/migrating-data-to-and-from-your-enterprise/exporting-migration-data-from-your-enterprise) with `ghe-migrator`
2. Upload them to Azure blob storage of your choice, in the preferred infrastructure and network.
3. Create pre-signed URLs to the archive using Azure's User Delegation SAS. 
4. Use the GEI CLI to actually start the migration, importing the archive into GitHub.com from the above pre-signed URL that we would pass to GEI CLI. This would entail using the command: `--git-archive-url GIT_SOURCE_ARCHIVE_URL --metadata-archive-url METADATA_ARCHIVE_URL`

It would be great to have your opinion on the above, as this would unblock the migration path. 


I also want to add more color to the motivation behind this feature request: When placing highly restricted sensitive code in storage accessible from the Internet, the security constraints imposed by large enterprises are typically tight - firewalls/IP allowlists are often a must, and using a shared key as an authentication method is deemed risky. 

Thank you for clarifying!

**Comment by timrogers** [comment]
@danielmeppiel Yes - that should work, as long as the end of result of step 3 is a URL which GitHub can access.
